docs/guides/implementation-guides/how-to-implement-dependency-injection.md

```diff
@@ -7,9 +7,120 @@ import MvvmArchitectureDiagram from '/img/gitbook-import/assets/image (3) (1) (2
 
 # How To Implement Dependency Injection
 
-Content in preparation.
+:::info
+For views you can make a simple [ViewLocator](../../concepts/view-locator) which will automatically resolve the view from the Locator.
+:::
 
 This guide will show you how to use Dependency Injection (DI) with _Avalonia UI_ and the MVVM pattern. 
 
 <img src={MvvmArchitectureDiagram} alt=""/>
 
+Let's assume that your MainViewModel as a dependency on MyService. The contructor for ViewModel should looks like this:
+```csharp
+private readonly MyService _myService;
+
+public MapsViewModel(MyService myService)
+{
+    _myService = myService;
+}
+```
+These are the steps you need to do in order to resolve that dependencu using DI.
+
+## Step 1: Install the NuGet package Microsoft.Extensions.DependencyInjection
+```shell
+dotnet add package Microsoft.Extensions.DependencyInjection
+```
+
+## Step 2: ServiceCollectionExtensions 
+The following code is creating an extension for IServiceCollection. This is where you will register all your common dependencies. It's recomended to create that class in the shared project used by all avalonia target platform.
+
+```csharp
+public static class ServiceCollectionExtensions {
+    public static void AddCommonServices(this IServiceCollection collection) {
+        collection.AddTransient<MainViewModel>();
+        collection.AddTransient<MyService>();
+    }
+}
+```
+
+## Step 3: ServiceManager
+The following code create a ServiceManager class that is used to retrieve an instance of a given object.
+```csharp
+public static class ServiceManager {
+    public static IServiceProvider Locator { get; private set; }
+
+    public static void InitialiseLocatorFromCollection(IServiceCollection collection) {
+        Locator = collection.BuildServiceProvider();
+    }
+}
+``` 
+
+## Step 4: Modify Program.cs
+This should be modified for each target platform project. You can add specific platform dependency at this location too (see comment in code below)
+
+```csharp
+class Program
+{
+    // Initialization code. Don't use any Avalonia, third-party APIs or any
+    // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
+    // yet and stuff might break.
+    [STAThread]
+    public static void Main(string[] args)
+    {
+        var collection = new ServiceCollection();
+        collection.AddCommonServices();
+
+        //Register platform specific services here with the collection
```

Comment:
It would be better to show a more practical example of this instead of just a comment. I did the comment because I was somewhat lazy at the time.

Reply:
Indeed. Do you have service in mind that would be fit for the role? I haven't written a multi-platform app in Avalonia yet.

Reply:
Hmm, of the top of my head a simple thing that does not really mean anything would be to create a service that returns a number depending on the platform. EG 1 on Linux, 2 on Windows, etc. 

If you want a more practical thing then you could make a service that reads file permissions. .NET has different methods for reading Windows file permissions compared to Unix ones.  If you try to use one on the on the wrong platform it will throw an exception.  A simple `CanRead()` method which returns a value indicating if the program can read a file should be enough.

Reply:
I wonder if it's worth it to add all this in this small guide. At the moment it seems we have the core concept of DI without too much fluff around. 

As a compromise we could maybe extend that comment to add a bit more info?

Reply:
If I can share my thoughts, I'd like to see an example of how to set up services for different platforms. That's the main reason I picked Avalonia UI. Maybe you could make two examples...

Reply:
> [!TIP]
> This can also go into Avalonia.Samples and cross linked to the docs.

Reply:
Yes I think adding a more specific example for multi-platform DI in Samples would be better.

Here it's just a very simple guide for DI in Avalonia, which would have help me greatly a few weeks ago when I started with my desktop app. :D